istioctl/cmd/istioctl/verifyinstall.go

```diff
@@ -0,0 +1,330 @@
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package main
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+
+	"github.com/ghodss/yaml"
+	"github.com/spf13/cobra"
+	"go.uber.org/multierr"
+	"istio.io/istio/pkg/kube"
+	"istio.io/istio/pkg/log"
+	v1batch "k8s.io/api/batch/v1"
+	"k8s.io/api/extensions/v1beta1"
+	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	apiextensionsclient "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	kclient "k8s.io/client-go/kubernetes"
+)
+
+var (
+	installFileName string
+	resourceList    []*unstructured.Unstructured
+	kubeClient      kclient.Interface
+	extensionClient *apiextensionsclient.Clientset
+)
+
+var (
+	verifyInstallCmd = &cobra.Command{
+		Use:   "verify-install",
+		Short: "Validate Istio Installation Status",
+		Long: `
+		verify-install Validate Istio Installation Status
+`,
+		Example: `
+istioctl verify-install -f istio-demo.yaml
+`,
+		RunE: func(c *cobra.Command, _ []string) (err error) {
+			if err = validateInputFlags(); err != nil {
+				return err
+			}
+
+			var reader io.Reader
+			if installFileName == "-" {
+				reader = os.Stdin
+			} else {
+				var in *os.File
+				if in, err = os.Open(installFileName); err != nil {
+					return err
+				}
+				reader = in
+				defer func() {
+					if errClose := in.Close(); errClose != nil {
+						log.Errorf("Error: close file from %s, %s", installFileName, errClose)
+						if err == nil {
+							err = errClose
+						}
+					}
+				}()
+			}
+			err = buildKubeClient(kubeconfig)
+			if err != nil {
+				return err
+			}
+			err = buildResourceListFromInstallationFile(reader)
+			if err != nil {
+				return err
+			}
+			return loopThroughResources(c.OutOrStdout())
+		},
+	}
+)
+
+func init() {
+	experimentalCmd.AddCommand(verifyInstallCmd)
+
+	verifyInstallCmd.PersistentFlags().StringVarP(&installFileName, "filename", "f",
+		"", "Input Istio Installation File, eg: istio-demo.yaml")
+}
+
+func buildResourceListFromInstallationFile(in io.Reader) error {
+	data, err := ioutil.ReadAll(in)
+	if err != nil {
+		return err
+	}
+	resourceList, err = parseFile(data)
+	return err
+}
+
+func loopThroughResources(out io.Writer) error {
+	for _, r := range resourceList {
+		kind := r.GetKind()
+		name := r.GetName()
+		namespace := r.GetNamespace()
+		apiVersion := r.GetAPIVersion()
+		resourceFullName := name + "/" + namespace
+		log.Debugf("verifying installation status for  Istio resource %s", kind+" "+name+" "+namespace+" "+apiVersion)
+		switch {
+		case apiVersion == "v1":
```

Comment:
I think this can be vastly simplified:

- The most basic thing to check is for existence. This is pretty much what you're doing here. Correct me if I am wrong, but I don't think you need to make calls using the strongly-typed APIs, you should be able to simply use Unstructured to read many of the standard ones. This allows us to write a single client function that can simply check whether the required resource exists or not. This should handle all the cases here.
- For other resources that we know we want to validate better, you can add a registry of "handlers" (i.e. kind, apiGroup, apiVersion => handler function or interface) that contains the specialized logic for performing additional validation. For example, for deployments, we can check whether the associated pods are healthy/ready or not.

Reply:
Thanks Oz for this suggestion, it is reduce the codes a lot. For the ones that we want to validate better, use the strongly-typed APIs, eg: deployment, err := kubeClient.ExtensionsV1beta1().Deployments(namespace).Get(name, meta_v1.GetOptions{}) to better check it's status?